libr/anal/p/anal_loongarch_gnu.c

```diff
@@ -0,0 +1,1338 @@
+{
+  "zero", "ra", "tp", "sp", "a0", "a1", "a2", "a3",
+  "a4", "a5", "a6", "a7", "t0", "t1", "t2", "t3",
+  "t4", "t5", "t6", "t7", "t8", "x", "fp", "s0",
+  "s1", "s2", "s3", "s4", "s5", "s6", "s7", "s8",
+};
+
+//use bit[30:26] to cal hash index
+#define LA_INSN_HASH(insn) (((insn) & 0x7c000000) >> 26)
+
+/**
+ * sign_extend32 - sign extend a 32-bit value using specified bit as sign-bit
+ * @value: value to sign extend
+ * @index: 0 based bit index (0<=index<32) to sign bit
+ *
+ * This is safe to use for 16- and 8-bit types as well.
+ */
+static inline st32 sign_extend32(ut32 value, int index)
+{
+	ut8 shift = 31 - index;
+	return (st32)(value << shift) >> shift;
+}
+
+/**
+ * sign_extend64 - sign extend a 64-bit value using specified bit as sign-bit
+ * @value: value to sign extend
+ * @index: 0 based bit index (0<=index<64) to sign bit
+ */
+static inline st64 sign_extend64(ut64 value, int index)
+{
+	ut8 shift = 63 - index;
+	return (st64)(value << shift) >> shift;
+}
+
+#define UL(x)       (x##UL)
+#define ULL(x)      (x##ULL)
+#define LA_PFM PFMT64x
+
+#define GENMASK_ULL(h, l) \
+	(((~ULL(0)) - (ULL(1) << (l)) + 1) & \
+	 (~ULL(0) >> (64 - 1 - (h))))
+
+#define GET_BIT(op, h, l) (((op)&GENMASK_ULL(h,l))>>(l))
+
+#define LA_RD() loongarch_r_lp64_name[GET_BIT((opcode), 4, 0)]
+#define LA_RJ() loongarch_r_lp64_name[GET_BIT((opcode), 9, 5)]
+#define LA_RK() loongarch_r_lp64_name[GET_BIT((opcode), 14, 10)]
+
+#define I_I26(op) ((((op)>>10) & 0xffff) | (((op)&0x3ff)<<16))  //26bit number
+#define I_I21(op) ((((op)>>10) & 0xffff)|(((op)&0x1f)<<16))
+#define I_I20(op) (((op)>>5) & 0xfffff)
+#define I_I16(op) (((op)>>10) & 0xffff)
+#define I_I14(op) (((op)>>10) & 0x3fff)
+#define I_I12(op) (((op)>>10) & 0xfff)
+#define I_I6(op) (((op)>>10) & 0x3f)
+#define I_I5(op) (((op)>>10) & 0x1f)
+#define I_SA2(op) (((op)>>15) & 0x3)
+#define I_SA3(op) (((op)>>15) & 0x7)
+
+#define I_I26s2(op) (I_I26((op)) << 2)
+#define I_I21s2(op) (I_I21((op))<<2)
+#define I_I14s2(op) (I_I14((op))<<2)
+#define I_I16s2(op) (I_I16((op)) <<2)
+
+#define I12_SX(op) sign_extend64(I_I12((op)), 11)
+#define I16_SX(op) sign_extend64(I_I16((op)), 15)
+
+#define I16s2_SX(op) sign_extend64(I_I16s2((op)), 16+1)
+#define I14s2_SX(op) sign_extend64(I_I14s2((op)), 16+1)
+#define I21s2_SX(op) sign_extend64(I_I21s2((op)), 21+1)
+#define I26s2_SX(op) sign_extend64(I_I26s2((op)), 26+1)
+
+#define ES_SX32(x) "32,"x",~"
+#define ES_B(x) "0xff,"x",&"
+#define ES_H(x) "0xffff,"x",&"
+#define ES_W(x) "0xffffffff,"x",&"
+#define ES_WH(x) "32,0xffffffff00000000,"x",&,>>"
+
+struct loongarch_opcode
+{
+  const ut32 match;
+  const ut32 mask; /* High 1 byte is main opcode and it must be 0xf. */
+  const char * const name;
+  const ut32 index;
+  const ut32 r_type; /*R_ANAL_OP_TYPE*/
+};
+
+#define HT_NUM 32
+struct loongarch_ASE
+{
+	struct loongarch_opcode *const opcode;
+	struct loongarch_opcode *la_opcode_ht[HT_NUM];
+	ut8 opc_htab_inited;
+
+};
+
+typedef enum la_insn {
```

Comment:
maybe move all that stuff into a .h file?. removing the asm plugin and just having the anal one will simplify duplicated symbols across plugins and fix some linking issues (as well as reduce build size)